File: dart/lib/src/trinsic_service.dart
import 'package:trinsic_dart/src/connect_service.dart';
import 'package:trinsic_dart/src/service_base.dart';

import 'package:trinsic_dart/trinsic.dart';

class TrinsicService extends ServiceBase {
  AccessManagementService? _accessManagementService;
  ConnectService? _connectService;
  CredentialService? _credentialService;
  FileManagementService? _fileManagementService;
  ProviderService? _providerService;
  TemplateService? _templateService;
  TrustRegistryService? _trustRegistryService;
  WalletService? _walletService;
  TrinsicService(TrinsicOptions? serverOptions) : super(serverOptions);

  AccessManagementService accessManagement() {
    _accessManagementService ??= AccessManagementService(serviceOptions);
    _accessManagementService!.serviceOptions = serviceOptions;
    return _accessManagementService!;
  }

  ConnectService connect() {
    _connectService ??= ConnectService(serviceOptions);
    _connectService!.serviceOptions = serviceOptions;
    return _connectService!;
  }

  CredentialService credential() {
    _credentialService ??= CredentialService(serviceOptions);
    _credentialService!.serviceOptions = serviceOptions;
    return _credentialService!;
  }

  FileManagementService fileManagement() {
    _fileManagementService ??= FileManagementService(serviceOptions);
    _fileManagementService!.serviceOptions = serviceOptions;
    return _fileManagementService!;
  }

  ProviderService provider() {
    _providerService ??= ProviderService(serviceOptions);
    _providerService!.serviceOptions = serviceOptions;
    return _providerService!;
  }

  TemplateService template() {
    _templateService ??= TemplateService(serviceOptions);
    _templateService!.serviceOptions = serviceOptions;
    return _templateService!;
  }

  TrustRegistryService trustRegistry() {
    _trustRegistryService ??= TrustRegistryService(serviceOptions);
    _trustRegistryService!.serviceOptions = serviceOptions;
    return _trustRegistryService!;
  }

  WalletService wallet() {
    _walletService ??= WalletService(serviceOptions);
    _walletService!.serviceOptions = serviceOptions;
    return _walletService!;
  }
}

File: web/src/TrinsicService.ts
import ServiceBase from "./ServiceBase";
import { TrinsicOptions } from "./proto/sdk/options/v1/options";
import { CredentialService } from "./CredentialService";
import { TemplateService } from "./TemplateService";
import { ProviderService } from "./ProviderService";
import { TrustRegistryService } from "./TrustRegistryService";
import { WalletService } from "./WalletService";
import { AccessManagementService } from "./AccessManagementService";
import { FileManagementService } from "./FileManagementService";

export class TrinsicService extends ServiceBase {
    private _access: AccessManagementService | undefined;
    private _credential: CredentialService | undefined;
    private _fileManagement: FileManagementService | undefined;
    private _provider: ProviderService | undefined;
    private _template: TemplateService | undefined;
    private _trustRegistry: TrustRegistryService | undefined;
    private _wallet: WalletService | undefined;

    constructor(options?: TrinsicOptions) {
        super(options);
    }

    public access(): AccessManagementService {
        this._access =
            this._access || new AccessManagementService(this.options);
        this._access.options = this.options;
        return this._access!;
    }

    public credential(): CredentialService {
        this._credential =
            this._credential || new CredentialService(this.options);
        this._credential.options = this.options;
        return this._credential!;
    }

    public fileManagement(): FileManagementService {
        this._fileManagement =
            this._fileManagement || new FileManagementService(this.options);
        this._fileManagement.options = this.options;
        return this._fileManagement!;
    }

    public provider(): ProviderService {
        this._provider = this._provider || new ProviderService(this.options);
        this._provider.options = this.options;
        return this._provider!;
    }

    public template(): TemplateService {
        this._template = this._template || new TemplateService(this.options);
        this._template.options = this.options;
        return this._template!;
    }

    public trustRegistry(): TrustRegistryService {
        this._trustRegistry =
            this._trustRegistry || new TrustRegistryService(this.options);
        this._trustRegistry.options = this.options;
        return this._trustRegistry!;
    }

    public wallet(): WalletService {
        this._wallet = this._wallet || new WalletService(this.options);
        this._wallet.options = this.options;
        return this._wallet!;
    }
}
